File: akka-docs/src/test/java/jdocs/stream/operators/flow/StatefulMap.java
/*
 * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.operators.flow;

import akka.actor.ActorSystem;
import akka.japi.Pair;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Source;

import java.util.*;
import java.util.stream.IntStream;

public class StatefulMap {
  static final ActorSystem system = null;

  public void indexed() {
    // #zipWithIndex
    Source.from(Arrays.asList("A", "B", "C", "D"))
        .statefulMap(
            () -> 0L,
            (index, element) -> Pair.create(index + 1, Pair.create(element, index)),
            indexOnComplete -> Optional.empty())
        .runForeach(System.out::println, system);
    // prints
    // Pair(A,0)
    // Pair(B,1)
    // Pair(C,2)
    // Pair(D,3)
    // #zipWithIndex
  }

  public void bufferUntilChanged() {
    // #bufferUntilChanged
    Source.from(Arrays.asList("A", "B", "B", "C", "C", "C", "D"))
        .statefulMap(
            () -> (List<String>) new LinkedList<String>(),
            (buffer, element) -> {
              if (buffer.size() > 0 && (!buffer.get(0).equals(element))) {
                return Pair.create(
                    new LinkedList<>(Collections.singletonList(element)),
                    Collections.unmodifiableList(buffer));
              } else {
                buffer.add(element);
                return Pair.create(buffer, Collections.<String>emptyList());
              }
            },
            Optional::ofNullable)
        .filterNot(List::isEmpty)
        .runForeach(System.out::println, system);
    // prints
    // [A]
    // [B, B]
    // [C, C, C]
    // [D]
    // #bufferUntilChanged
  }

  public void distinctUntilChanged() {
    // #distinctUntilChanged
    Source.from(Arrays.asList("A", "B", "B", "C", "C", "C", "D"))
        .statefulMap(
            Optional::<String>empty,
            (lastElement, element) -> {
              if (lastElement.isPresent() && lastElement.get().equals(element)) {
                return Pair.create(lastElement, Optional.<String>empty());
              } else {
                return Pair.create(Optional.of(element), Optional.of(element));
              }
            },
            listOnComplete -> Optional.empty())
        .via(Flow.flattenOptional())
        .runForeach(System.out::println, system);
    // prints
    // A
    // B
    // C
    // D
    // #distinctUntilChanged
  }

  public void statefulMapConcatLike() {
    // #statefulMapConcatLike
    Source.fromJavaStream(() -> IntStream.rangeClosed(1, 10))
        .statefulMap(
            () -> new ArrayList<Integer>(3),
            (list, element) -> {
              list.add(element);
              if (list.size() == 3) {
                return Pair.create(new ArrayList<Integer>(3), Collections.unmodifiableList(list));
              } else {
                return Pair.create(list, Collections.<Integer>emptyList());
              }
            },
            Optional::ofNullable)
        .mapConcat(list -> list)
        .runForeach(System.out::println, system);
    // prints
    // 1
    // 2
    // 3
    // 4
    // 5
    // 6
    // 7
    // 8
    // 9
    // 10
    // #statefulMapConcatLike
  }
}

File: akka-docs/src/test/scala/docs/stream/operators/flow/StatefulMap.scala
/*
 * Copyright (C) 2022-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.operators.flow
import akka.actor.ActorSystem
import akka.stream.scaladsl.Source

object StatefulMap {

  implicit val actorSystem: ActorSystem = ???

  def indexed(): Unit = {
    //#zipWithIndex
    Source(List("A", "B", "C", "D"))
      .statefulMap(() => 0L)((index, elem) => (index + 1, (elem, index)), _ => None)
      .runForeach(println)
    //prints
    //(A,0)
    //(B,1)
    //(C,2)
    //(D,3)
    //#zipWithIndex
  }

  def bufferUntilChanged(): Unit = {
    //#bufferUntilChanged
    Source("A" :: "B" :: "B" :: "C" :: "C" :: "C" :: "D" :: Nil)
      .statefulMap(() => List.empty[String])(
        (buffer, element) =>
          buffer match {
            case head :: _ if head != element => (element :: Nil, buffer)
            case _                            => (element :: buffer, Nil)
          },
        buffer => Some(buffer))
      .filter(_.nonEmpty)
      .runForeach(println)
    //prints
    //List(A)
    //List(B, B)
    //List(C, C, C)
    //List(D)
    //#bufferUntilChanged
  }

  def distinctUntilChanged(): Unit = {
    //#distinctUntilChanged
    Source("A" :: "B" :: "B" :: "C" :: "C" :: "C" :: "D" :: Nil)
      .statefulMap(() => Option.empty[String])(
        (lastElement, elem) =>
          lastElement match {
            case Some(head) if head == elem => (Some(elem), None)
            case _                          => (Some(elem), Some(elem))
          },
        _ => None)
      .collect { case Some(elem) => elem }
      .runForeach(println)
    //prints
    //A
    //B
    //C
    //D
    //#distinctUntilChanged
  }

  def statefulMapConcatLike(): Unit = {
    //#statefulMapConcatLike
    Source(1 to 10)
      .statefulMap(() => List.empty[Int])(
        (state, elem) => {
          //grouped 3 elements into a list
          val newState = elem :: state
          if (newState.size == 3)
            (Nil, newState.reverse)
          else
            (newState, Nil)
        },
        state => Some(state.reverse))
      .mapConcat(identity)
      .runForeach(println)
    //prints
    //1
    //2
    //3
    //4
    //5
    //6
    //7
    //8
    //9
    //10
    //#statefulMapConcatLike
  }
}
